File: 40 - Combination Sum II/solution.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {

    public void helper(int[] nums, int target, int ind, List<Integer> arr, List<List<Integer>> result){
        if(target == 0){
            result.add(new ArrayList<>(arr));
            return;
        }

        for(int i=ind; i<nums.length; i++){
            if(i>ind && nums[i] == nums[i-1]) continue;
            if(nums[i] > target) break;

            arr.add(nums[i]);
            helper(nums, target-nums[i], i+1, arr, result);
            arr.remove(arr.size()-1);
        }
    }

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();

        Arrays.sort(candidates);

        helper(candidates, target, 0, new ArrayList<>(), result);

        return result;
    }
}

File: 40 - Combination Sum II/solution.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
    candidates.sort((a,b) => a-b);
    const result = [];

    const helper = (ind, target, arr) => {
        if(target == 0){
            result.push([...arr]);
            return;
        }

        for(let i=ind; i<candidates.length; i++){
            if(i>ind && candidates[i]==candidates[i-1]) continue;
            if(candidates[i] > target) break;

            arr.push(candidates[i]);
            helper(i+1, target-candidates[i], arr);
            arr.pop();
        }
    }

    helper(0, target, []);
    return result;
};